src/betterproto/__init__.py

```diff
@@ -469,22 +551,33 @@ def _dump_float(value: float) -> Union[float, str]:
     return value
 
 
-def decode_varint(buffer: bytes, pos: int) -> Tuple[int, int]:
+def load_varint(stream: BinaryIO) -> Tuple[int, bytes]:
     """
-    Decode a single varint value from a byte buffer. Returns the value and the
-    new position in the buffer.
+    Load a single varint value from a stream. Returns the value and the raw bytes read.
     """
     result = 0
-    shift = 0
-    while 1:
-        b = buffer[pos]
-        result |= (b & 0x7F) << shift
-        pos += 1
-        if not (b & 0x80):
-            return result, pos
-        shift += 7
+    raw = b""
+    for shift in count(0, 7):
         if shift >= 64:
             raise ValueError("Too many bytes when decoding varint.")
+        b = stream.read(1)
+        if b == b"":
+            raise EOFError("Stream ended unexpectedly while attempting to load varint.")
+        raw += b
+        b_int = int.from_bytes(b)
+        result |= (b_int & 0x7F) << shift
+        if not (b_int & 0x80):
+            return result, raw
+
+
+def decode_varint(buffer: bytes, pos: int) -> Tuple[int, int]:
+    """
+    Decode a single varint value from a byte buffer. Returns the value and the
+    new position in the buffer.
+    """
+    with BytesIO(buffer[pos:]) as stream:
```

Comment:
You can avoid a potential copy of the bytes via the slice with:

```suggestion
    with BytesIO(buffer) as stream:
        stream.seek(pos)
```

Reply:
Changed